src/frequenz/sdk/actor/_data_sourcing/microgrid_api_source.py

```diff
@@ -343,18 +343,39 @@ async def _handle_data_stream(
             stream_senders = self._get_metric_senders(
                 category, self._req_streaming_metrics[comp_id]
             )
-        api_data_receiver = self.comp_data_receivers[comp_id]
+        api_data_receiver: Receiver[Any] = self.comp_data_receivers[comp_id]
+
+        senders_done: asyncio.Event = asyncio.Event()
+        pending_messages = 0
 
         def process_msg(data: Any) -> None:
             tasks = []
             for extractor, senders in stream_senders:
                 for sender in senders:
                     tasks.append(sender.send(Sample(data.timestamp, extractor(data))))
             asyncio.gather(*tasks)
+            nonlocal pending_messages
+            pending_messages -= 1
+            if pending_messages == 0:
+                senders_done.set()
 
         async for data in api_data_receiver:
+            pending_messages += 1
+            senders_done.clear()
             process_msg(data)
 
+        while pending_messages > 0:
+            await senders_done.wait()
```

Comment:
I won't argue but to remove all this complexity we could do: 
```
 async for data in api_data_receiver:
       await process_msg(data)
```

This is also much save because:
* we will send data in the order (everywhere in the code we assume it is correct order)
*  api_data_receiver has maxsize =  50. If we exceed this number, then we start to see logs that messages get lost and we are not processing them in time. 
* In current solution we won't see any logs, we will just produce infinite number of async task, that would block everything.

Reply:
you mean we await the `asyncio.gather(*tasks)` call? In that case I don't need all the complexity here.. But didn't we want to wait for sahas before making such changes? :)

Reply:
I tried to keep the impact on the code behavior low because of that

Reply:
Sounds good to me. I think whatever we do now is likely to change if we go for a sync `send()`. We are just speculating now.